src/sql-parser/src/parser.rs

```diff
@@ -566,13 +566,37 @@ impl<'a> Parser<'a> {
             Token::Keyword(EXTRACT) => self.parse_extract_expr(),
             Token::Keyword(INTERVAL) => Ok(Expr::Value(self.parse_interval_value()?)),
             // having timestamps separately here because errors are swallowed by the maybe! block
-            Token::Keyword(TIMESTAMP) | Token::Keyword(TIMESTAMPTZ) => {
-                self.prev_token();
-                let data_type = self.parse_data_type()?;
-                Ok(Expr::Cast {
-                    expr: Box::new(Expr::Value(Value::String(self.parse_literal_string()?))),
-                    data_type,
-                })
+            Token::Keyword(id @ TIMESTAMP) | Token::Keyword(id @ TIMESTAMPTZ) => {
+                let index = self.index;
+                let maybe_data_type = |parser: &mut Parser| {
+                    parser.prev_token();
+                    match parser.parse_data_type() {
+                        Ok(data_type) => Ok(Expr::Cast {
+                            expr: Box::new(Expr::Value(Value::String(
+                                parser.parse_literal_string()?,
+                            ))),
+                            data_type,
+                        }),
+                        Err(e) => Err(e),
+                    }
+                };
+
+                // Check if it's a valid data type.
+                // If not, check if it can be a identifier.
+                // Return the data type parsing error in case all fails
+                match maybe_data_type(self) {
+                    Ok(data_type_expr) => Ok(data_type_expr),
+                    Err(e) => {
+                        self.index = index;
+                        match self.parse_qualified_identifier(id.into()) {
+                            // It cannot be a function, it's a type with modifiers
+                            // Returning the original parsing error
+                            Ok(Expr::Function(_)) => Err(e),
+                            Ok(expr) => Ok(expr),
+                            Err(_) => Err(e),
+                        }
+                    }
+                }
```

Comment:
Ugh, I see what you mean about it being hard to adapt `maybe_parse` for this. 

However, this logic is suuuuper subtle. I'm having trouble reasoning about it. It seems like a lot of complexity just to improve that one error message.

What do you think about doing a more targeted fix for this release that just removes this entire special case (and intentionally regresses the error message), and then we can experiment with some alternative methods of recovering the good error message without the time pressure.

Thanks so much for working on this one!

Reply:
> What do you think about doing a more targeted fix for this release that just removes this entire special case (and intentionally regresses the error message), and then we can experiment with some alternative methods of recovering the good error message without the time pressure.

Personally I would also prefer this, thank you @moulimukherjee for diving deep here!

Reply:
Sounds good 👍